### PR TITLE
Abandon ship

### PR DESCRIPTION
**Pull request by mkerensky**
+Bugfix  Correctly applies autoload of transported units in lobby on a per-ship basis
+Feature  Crews of large craft and units transported aboard large craft are handled during scenario resolution

### CODE REVIEW

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 1]
```diff
@@ -395,7 +409,7 @@ private void configureBot(BotClient botClient, BotForce botForce) {
                 botClient.getLocalPlayer().setTeam(botForce.getTeam());
                 botClient.getLocalPlayer().setStartingPos(botForce.getStart());
 
-                if (botForce.getCamoCategory().equals(Player.NO_CAMO)) {
+                if (botForce.getCamoCategory() == Player.NO_CAMO) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]: Why was this swapped from .equals? That makes no sense

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 2]
```diff
@@ -634,6 +673,180 @@ else if (pilot.getSlotCount() > 1) {
...
+                        //Let's assume people are injured on an 8+ if the unit is destroyed, same as a critical hit chance
+                        if (e.isDestroyed() && Compute.d6(2) >= 8) {
+                            //As with crewmembers, on a 7+ they're only wounded
+                            if(Compute.d6(2) >= 7) {
+                                wounded = true;
+                            } else {
+                                status.setHits(6);
+                                status.setDead(true);
+                            }
+                            if(wounded) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                            if (wounded) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 3]
```diff
@@ -634,6 +673,180 @@ else if (pilot.getSlotCount() > 1) {
...
+                    cargoCrew = shuffleCrew(cargoCrew);
+                    for (Person p : cargoCrew) {
+                        PersonStatus status = new PersonStatus(p.getFullName(), u.getEntity().getDisplayName(),
+                                p.getHits(), p.getId());
+                        boolean wounded = false;
+                        //The lore says bay crews have pressurized sleeping alcoves in the corners of each bay
+                        //Let's assume people are injured on an 8+ if the unit is destroyed, same as a critical hit chance
+                        if (e.isDestroyed() && Compute.d6(2) >= 8) {
+                            //As with crewmembers, on a 7+ they're only wounded
+                            if(Compute.d6(2) >= 7) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                            if (Compute.d6(2) >= 7) {
```

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 4]
```diff
@@ -117,14 +117,17 @@
     // Convenience data used by GameThread
     private boolean carryingAero = false;
     private boolean carryingGround = false;
-
+
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java]:
```suggestion

```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 5]
```diff
@@ -51,7 +51,7 @@
 
 /**
  * @author Neoancient
- *
+ *
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
 *
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 6]
```diff
@@ -128,7 +128,7 @@ public void run() {
                 MapSettings mapSettings = MapSettings.getInstance();
 
                 // if the scenario is taking place in space, do space settings instead
-                if ((scenario instanceof AtBScenario) &&
+                if((scenario instanceof AtBScenario) &&
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
                if ((scenario instanceof AtBScenario) &&
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 7]
```diff
@@ -153,7 +153,7 @@ public void run() {
                 client.sendMapSettings(mapSettings);
                 Thread.sleep(campaign.getCampaignOptions().getStartGameDelay());
 
-                PlanetaryConditions planetaryConditions = new PlanetaryConditions();
+                PlanetaryConditions planetaryConditions = new PlanetaryConditions();
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
                PlanetaryConditions planetaryConditions = new PlanetaryConditions();
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 8]
```diff
@@ -223,7 +229,7 @@ public void run() {
                     entity.setDeployRound(deploymentRound);
                     // Add Mek to game
                     client.sendAddEntity(entity);
-                    // Wait a few secs to not overuse bandwidth
+                    // Wait a few secs to not overuse bandwith
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
                    // Wait a few secs to not overuse bandwidth
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 9]
```diff
@@ -199,6 +199,12 @@ public void run() {
                         //Store this unit as a potential transport to load
                         scenario.getPlayerTransportLinkages().put(unit.getId(), new ArrayList<>());
                     }
+                    // If this unit is a spacecraft, set the crew size and marine size values
+                    if (entity.isLargeCraft() || entity.getUnitType() == UnitType.SMALL_CRAFT) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
                    if (entity.isLargeCraft() || (entity.getUnitType() == UnitType.SMALL_CRAFT)) {
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 10]
```diff
@@ -408,6 +422,6 @@ private void configureBot(BotClient botClient, BotForce botForce) {
             }
         } catch (Exception e) {
             MekHQ.getLogger().error(getClass(), "configureBot", e); //$NON-NLS-1$
-        }
+        }
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
        }
```

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 1 above)
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
                if (botForce.getCamoCategory().equals(Player.NO_CAMO)) {
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 11]
```diff
@@ -367,7 +381,7 @@ public void run() {
     /**
      * wait for the server to add the bot client, then send starting position,
      * camo, and entities
-     *
+     *
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]:
```suggestion
     *
```

File: MekHQ/src/mekhq/AtBGameThread.java [hunk 12]
```diff
@@ -317,33 +323,41 @@ public void run() {
...
+                        }
+                        if (transport.isCarryingGround()) {
+                            loadGround = (JOptionPane.YES_OPTION == JOptionPane.showConfirmDialog(null,
+                                    String.format(AtBGameThread.LOAD_GND_DIALOG_TEXT, transport.getName()),
+                                    AtBGameThread.LOAD_GND_DIALOG_TITLE, JOptionPane.YES_NO_OPTION));
+                        }
+                        // Now, send the load commands
+                        if (loadFighters || loadGround) {
+                            // List of technicians assigned to transported units. Several units can share a tech.
+                            Set<UUID> cargoTechs = new HashSet<>();
```
Comment (Windchild292) [inline MekHQ/src/mekhq/AtBGameThread.java]: I'm assuming this is to ensure the UUIDs are all unique?

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 13]
```diff
@@ -375,7 +412,7 @@ private UnitStatus processAlliedUnit(Entity e) {
         UnitStatus us = new UnitStatus(nu);
         unitsStatus.put(nu.getId(), us);
         alliedUnits.add(nu);
-
+
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion

```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 14]
```diff
@@ -473,158 +510,160 @@ public void checkStatusOfPersonnel() {
...
-                    existingHits = u.getEntity().getCrew().getHits();
+                processLargeCraft(u, en, crew, unitStatus);
+            } else {
+                if (en.getTransportId() != Entity.NONE) {
+                    // Check to see if the unit is in a large craft bay, if so, its crew will be processed with the ship,
+                    // so ignore it here.
+                    UUID trnId = idMap.get(en.getTransportId());
+                    if (trnId != null) {
+                        Entity transport = unitsStatus.get(trnId).getEntity();
+                        if (transport != null && transport.isLargeCraft()) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                        if ((transport != null) && transport.isLargeCraft()) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 15]
```diff
@@ -473,158 +510,160 @@ public void checkStatusOfPersonnel() {
...
-                                continue;
-                            }
-                            if (en.getInternal(loc) <= 0) {
-                                destroyed = true;
+                }
+                if (null == pilot) {
+                    pilot = mia.get(UUID.fromString(en.getCrew().getExternalIdAsString()));
+                    missingCrew = true;
+                }
+                for(Person p : crew) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                for (Person p : crew) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 16]
```diff
@@ -473,158 +510,160 @@ public void checkStatusOfPersonnel() {
...
-                                status.setDead(true);
+                            if (destroyed || null == en.getCrew() || en.getCrew().isDead()) {
+                                if (Compute.d6(2) >= 7) {
+                                    wounded = true;
+                                } else {
+                                    status.setHits(6);
+                                    status.setDead(true);
+                                }
+                            } else if (((Tank)en).isDriverHit() && u.isDriver(p)) {
+                                if(Compute.d6(2) >= 7) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                                if (Compute.d6(2) >= 7) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 17]
```diff
@@ -473,158 +510,160 @@ public void checkStatusOfPersonnel() {
...
+                                    wounded = true;
+                                } else {
+                                    status.setHits(6);
+                                    status.setDead(true);
+                                }
+                            } else if (((Tank)en).isCommanderHit() && (u.isCommander(p)
+                                    || u.isTechOfficer(p))) {
+                                //If there is a command console, the commander hit flag is set on the second such critical,
+                                //which means both commanders have been hit.
+                                if(Compute.d6(2) >= 7) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                                if (Compute.d6(2) >= 7) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 18]
```diff
@@ -473,158 +510,160 @@ public void checkStatusOfPersonnel() {
...
+                                    wounded = true;
+                                } else {
+                                    status.setHits(6);
+                                    status.setDead(true);
+                                }
+                            } else if (((Tank)en).isUsingConsoleCommander() && u.isCommander(p)) {
+                                //This flag is set after the first commander hit critical.
+                                if(Compute.d6(2) >= 7) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                                if (Compute.d6(2) >= 7) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 19]
```diff
@@ -473,158 +510,160 @@ public void checkStatusOfPersonnel() {
...
+                                    wounded = true;
+                                } else {
+                                    status.setHits(6);
+                                    status.setDead(true);
+                                }
                             }
-                        } else if (((Tank)en).isUsingConsoleCommander() && u.isCommander(p)) {
-                            //This flag is set after the first commander hit critical.
+                        }
+                        if(casualtiesAssigned < casualties) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                        if (casualtiesAssigned < casualties) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 20]
```diff
@@ -473,158 +510,160 @@ public void checkStatusOfPersonnel() {
...
-                            status.setHits(6);
-                            status.setDead(true);
-                        }
-                    }
-                    if(wounded) {
-                        int hits = campaign.getCampaignOptions().getMinimumHitsForVees();
-                        if (campaign.getCampaignOptions().useAdvancedMedical() || campaign.getCampaignOptions().useRandomHitsForVees()) {
-                            int range = 6 - hits;
-                            hits = hits + Compute.randomInt(range);
+                        if(wounded) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                        if (wounded) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 21]
```diff
@@ -634,6 +673,180 @@ else if (pilot.getSlotCount() > 1) {
...
+        casualties = (int) Math.ceil(Compute.getFullCrewSize(en) * (newHits / 6.0));
+        //Now reduce the casualties if some "hits" were caused by ejection
+        casualties = Math.max(0, casualties - rescuedCrew);
+        
+        //And assign the casualties and experience amongst the crew and marines
+        for(Person p : personnel) {
+            PersonStatus status = new PersonStatus(p.getFullName(), ship.getEntity().getDisplayName(),
+                    p.getHits(), p.getId());
+            boolean wounded = false;
+            if(casualtiesAssigned < casualties) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
            if (casualtiesAssigned < casualties) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 22]
```diff
@@ -634,6 +673,180 @@ else if (pilot.getSlotCount() > 1) {
...
+        if (en.isDestroyed()) {
+            currentHits = 6;
+        }
+        int newHits = Math.max(0, currentHits - existingHits);
+        casualties = (int) Math.ceil(Compute.getFullCrewSize(en) * (newHits / 6.0));
+        //Now reduce the casualties if some "hits" were caused by ejection
+        casualties = Math.max(0, casualties - rescuedCrew);
+        
+        //And assign the casualties and experience amongst the crew and marines
+        for(Person p : personnel) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
        for (Person p : personnel) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 23]
```diff
@@ -634,6 +673,180 @@ else if (pilot.getSlotCount() > 1) {
...
+            PersonStatus status = new PersonStatus(p.getFullName(), ship.getEntity().getDisplayName(),
+                    p.getHits(), p.getId());
+            boolean wounded = false;
+            if(casualtiesAssigned < casualties) {
+                casualtiesAssigned++;
+                if(Compute.d6(2) >= 7) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
                if (Compute.d6(2) >= 7) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 24]
```diff
@@ -634,6 +673,180 @@ else if (pilot.getSlotCount() > 1) {
...
+                    wounded = true;
+                } else {
+                    status.setHits(6);
+                    status.setDead(true);
+                }
+            }
+            if(wounded) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
            if (wounded) {
```

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java [hunk 25]
```diff
@@ -634,6 +673,180 @@ else if (pilot.getSlotCount() > 1) {
...
+                if (campaign.getCampaignOptions().useAdvancedMedical() || campaign.getCampaignOptions().useRandomHitsForVees()) {
+                    int range = 6 - hits;
+                    hits = hits + Compute.randomInt(range);
+                }
+                status.setHits(hits);
+            }
+        status.setXP(campaign.getCampaignOptions().getScenarioXP());
+        status.setDeployed(!en.wasNeverDeployed());
+        peopleStatus.put(p.getId(), status);
+        }
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java]:
```suggestion
            }
            status.setXP(campaign.getCampaignOptions().getScenarioXP());
            status.setDeployed(!en.wasNeverDeployed());
            peopleStatus.put(p.getId(), status);
        }
```

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 5 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:54]: Fixed

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 6 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:131]: Fixed

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 7 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:156]: Fixed

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 9 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:203]: Fixed

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 8 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:232]: Fixed

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 12 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:350]: Correct.  You can have the same tech assigned to multiple units, and I only want one person to be added aboard the transport in such cases.

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 11 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:384]: Fixed

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 1 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:412]: Couldn't say. I suspect it was introduced from a merge of the latest Master.

File: MekHQ/src/mekhq/AtBGameThread.java (hunk 10 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/AtBGameThread.java:425]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 13 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:415]: Fixed. My IDE puts this space in when I add a new line. I think I fixed that in preferences from here forward.

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 14 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:523]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 15 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:574]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 16 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:619]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 17 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:629]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 18 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:637]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 19 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:645]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 20 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:654]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 22 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:745]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 21 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:749]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 23 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:751]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 24 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:758]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 25 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:769]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 3 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:791]: Fixed

File: MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java (hunk 2 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/ResolveScenarioTracker.java:797]: Fixed

File: MekHQ/src/mekhq/campaign/unit/Unit.java (hunk 4 above)
Comment (mkerensky) [reply MekHQ/src/mekhq/campaign/unit/Unit.java:120]: Fixed

**Comment by NickAragua** [comment]
I'll get this merged once we take care of the megamek PR.

**Pull request merged**